### ISSUE TITLE
Feature request: TaskPaper-style editor

### ISSUE THREAD

**Issue by triangulum** [sn-extensions/simple-task-editor#8]
The simple task editor works well for *simple* task lists (hence the name!). Yet I find myself missing the functionality of other plain text task managers, such as [TaskPaper](https://www.taskpaper.com/). This could also be thought of as a simplified digital [Bullet Journal](https://bulletjournal.com/).

![](https://i.imgur.com/gtR77lo.gif)

Has anyone on the development team used TaskPaper before? Would it be possible to consider a "TaskPaper" extension, as an advanced-task-editor for Standard Notes? It seems like the functionality of TaskPaper naturally lends itself to the cross-platform and secure syncing nature of Standard Notes!

Things that I (think I) am interested in (not a definitive list!):
- plain-text backend with keyboard navigation and shortcuts
- project headings, each with their own list (and sub-lists)
- _notes_
- completed tasks maintain their position in the list (until archived)
- potentially: `@tags`

There's a nice [User Guide](https://guide.taskpaper.com/getting-started/) that outlines most of the functionality. It's also worth noting that the format seems to be fairly stable, and has been integrated by many other applications (which is good for longevity)!

---

There are also interesting discussions for interfacing with the TaskPaper format on macOS and iOS:
- What other apps support TaskPaper’s file format? https://support.hogbaysoftware.com/t/what-other-apps-support-taskpapers-file-format/1114
- Extending the iOS Drafts app into a TaskPaper client: https://leancrew.com/all-this/2018/05/taskpaper-actions-for-drafts-5/
- Editorial support for the TaskPaper format: http://omz-software.com/editorial/docs/ios/editorial_writing_todo_lists.html
- OmniFocus TaskPaper reference https://support.omnigroup.com/omnifocus-taskpaper-reference/

**Comment by bryvin** [comment]
I've never heard of that one before, but it seems pretty standard as far as task lists go.

I don't believe Mo will include this feature set in the current version of the simple-task-editor.. but certainly it could be developed by anyone as a separate Extension. Sounds like a fun project to even maybe learn some programming with if you're new to that scene. Definitely check out the [Intro to Extensions](https://docs.standardnotes.org/extensions/intro-to-extensions) documentation.

**Comment by mobitar** [comment]
Indeed, not something we'd include in Simple Task Editor. TaskPaper seems like an outline edtior, like workflowy? If so, we've looked into this before, and it's hard to develop from scratch, and open-source ones don't cut it. So unless someone else develops it for SN, it's not likely we'd allocate the several months it would require to build this.

**Comment by triangulum** [comment]
Kind of like workflowy, except that everything is stored as a plain text file. So that would really reduce the complexity to just parsing the file and displaying it properly.

The developer of TaskPaper even has an [open source model layer](https://support.hogbaysoftware.com/t/what-is-taskpapers-open-source-model-layer/2148?u=jessegrosjean), to make it easier to include the standard in other applications!

> TaskPaper’s open source model layer (Birch Outline, MIT License) makes it much easier for other developers and scripters to read, process, and write TaskPaper files.

> Birch Outline is a JavaScript package and also includes a Swift wrapper for macOS and iOS apps. It gives access to much of TaskPaper’s runtime behavior including: Read/Write TaskPaper formatted text, runtime model, and query evaluation.

> Do you have a favorite app that would be improved with TaskPaper support? If so please let that app’s developer know about TaskPaper’s new open source model layer. It makes that job of working with TaskPaper files much easier.

**Comment by mobitar** [comment]
It's really the UI that makes this difficult. I don't have the capacity to build this myself, but if anyone wants to volunteer, by all means!
